### ISSUE TITLE
Promotion voucher still applied even if condition has changed

### ISSUE THREAD

**Issue by pawel-0** [Sylius/Sylius#9361]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Sylius version   | 1.1.5

Steps to reproduct:
1. Add products to cart with high qty to reach item price > $100
2. Apply voucher (Condition: item total = $100; Discount = $10)
3. decrease qty of item so item price is < $100

Expexted Result:
- Discount is removed

Actual result:
- some kind of discount is still calculated

**Comment by Zales0123** [comment]
Indeed there is a problem! It happens because discount is distributed over item units, and even though coupon is no longer valid after changing quantity, the discount is still applied on remaining units. Thanks @pawelwaniek for reporting that 👍 

<img width="1172" alt="zrzut ekranu 2018-04-24 o 08 23 47" src="https://user-images.githubusercontent.com/6212718/39169679-05f4edfc-4799-11e8-83de-9ae50accf937.png">
<img width="1219" alt="zrzut ekranu 2018-04-24 o 08 23 42" src="https://user-images.githubusercontent.com/6212718/39169677-05d948d6-4799-11e8-9a6c-b8746f5fde84.png">

**Comment by pawelwaniek** [comment]
I think there is a general issue with the promotion recalculation if any condition changes.

Discounts are still applied after:
- disabling the promotion for this channel
- changing the start/end date to a invalid one
- changing the action e.g. fix discount

**Comment by pamil** [comment]
Is it possible to complete checkout process with that promotion?

**Comment by pawelwaniek** [comment]
If you change the qty (in the case I described) to a lower qty the page reloads and the grand total is recalculated and you get some leftover discount. But if you reload the cart (F5 or submit url again) you see the previous qty again and the old grand total, so the first qty change is not saved permanently.
That means if you proceed the checkout you will see on the checkout "Review" Page the old qty with the correct price and discount

**Comment by pamil** [comment]
Related issues: #8320, #8699.

**Comment by jcgdjob** [comment]
Just to share how I "fix" it:
decorates `Validator/PromotionSubjectCouponValidator.php`

Set promotion coupon to null if not Eligible

```
if ($this->promotionEligibilityChecker->isEligible($value, $promotionCoupon->getPromotion())) {
    return;
} else {
    $value->setPromotionCoupon(null);
    $this->entityManager->flush();
    $this->orderProcessor->process($value);
    return;
}
```
on twig, add js to remove coupon code if error message returned.

**Comment by andriusvo** [comment]
⬆️

**Comment by jakubtobiasz** [comment]
Hello 👋🏼,
this issue has been resolved by #14873 & #14865.
